sympy/solvers/ode/ode.py

```diff
@@ -1752,6 +1752,209 @@ def match_2nd_linear_bessel(r, func):
+                    eq_no = i
+            if eq_no in func_dict:
+                list_func = []
+                list_func.append(func_dict[eq_no])
+                list_func.append(func)
+                func_dict[eq_no] = list_func
+            else:
+                func_dict[eq_no] = func
+            order[func] = max_order
+
+    funcs = [func_dict[i] for i in range(len(func_dict))]
+
+    return funcs, order
+
+
+def _get_independent_variable(eq):
+    return list(list(eq.atoms(Derivative))[0].atoms(Symbol))[0]
+
+
+def get_linear_coeffs(eq, funcs):
+    '''
+    This function returns the coefficients of the equation if it is linear
+    else it returns None.
+
+    Paramters
+    =========
+
+    eq: Equation
+        ODE for which the check of linearity is to be done.
+    funcs: List of variables
+        List of dependent variables for which the linearity of the eq has to
+        be checked
+
+    Returns
+    =======
+
+    Dict or None
+        If the eq is linear ODE, then it returns a dictionary where the keys
+        are the terms of ODE and values are its coefficients.
+        If the eq is not linear, then None is returned
+    '''
+    coeffs = dict()
+    t = _get_independent_variable(eq)
+    is_linear_ = True
+    for func in funcs:
+        order_func = ode_order(eq, func)
+        for k in range(order_func + 1):
+            coeffs[diff(func, t, k)] = collect(eq.expand(), [diff(func, t, k)])\
+                                        .coeff(diff(func, t, k))
+            if is_linear_ == True:
+                if coeffs[diff(func, t, k)] == 0:
+                    if k == 0:
+                        coef = eq.as_independent(func, as_Add=True)[1]
+                        for xr in range(1, ode_order(eq, func) + 1):
+                            coef -= eq.as_independent(diff(func, t, xr), as_Add=True)[1]
+                        if coef != 0:
+                            is_linear_ = False
+                    else:
+                        if eq.as_independent(diff(func, t, k), as_Add=True)[1]:
+                            is_linear_ = False
+                else:
+                    for func_ in funcs:
+                        if isinstance(func_, list):
+                            for elem_func_ in func_:
+                                dep = coeffs[diff(func, t, k)].as_independent(elem_func_, as_Add=True)[1]
+                                if dep != 0:
+                                    is_linear_ = False
+                        else:
+                            dep = coeffs[diff(func, t, k)].as_independent(func_, as_Add=True)[1]
+                            if dep != 0:
+                                is_linear_ = False
+    if is_linear_:
+        return coeffs
+    return None
+
+
+def neq_nth_linear_constant_coeff_match(eqs, funcs):
+    from sympy.solvers.solveset import linear_eq_to_matrix
+
+    match = dict()
+    match["no_of_equation"] = len(eqs)
+    match["eq"] = eqs
```

Comment:
It would be better to fill out the contents of match once at the end of the function e.g.:
```
match = {
    'no_of_equation': len(eqs),
    ...
}
```